test/e2e/e2e_test.go

```diff
@@ -42,8 +46,16 @@ import (
 	_ "k8s.io/kubernetes/test/e2e/ui"
 )
 
+var viperConfig = flag.String("viper-config", "", "The name of a viper config file (https://github.com/spf13/viper#what-is-viper). All e2e command line parameters can also be configured in such a file. May contain a path and may or may not contain the file suffix. The default is to look for an optional file with `e2e` as base name. If a file is specified explicitly, it must be present.")
```

Comment:
This sort of distributed registration makes it really hard to depend on packages because of conflicts.  It also makes it impossible to avoid stripping vendor.  And it prevents visibility to any other thing that wants access.  I don't think we should use a pattern like this and instead we could use the `kubectl` or `kube-apiserver/controller-manager` method of `Options` structs with `Bind` methods.

Reply:
The idea was to let each test own a certain parameter namespace (like 'storage' for test/e2e/storage). One of my earlier revisions of this PR described a naming scheme that would have avoided conflicts, but during review it was pointed out that this is too complex and therefore its now only a recommendation.

`UpgradeTarget` and `UpgradeImage` are legacy options that for historic reasons are defined at the top level of the namespace and weren't renamed to avoid breaking jobs. They don't belong into the core framework because they only parametrize certain tests.

Reply:
This has actually made it harder to build tests on top.  Instead of having a publicly exposed config struct that I can set any way I want and then use, I must have my executable called to set these registered flags.

Reply:
i see viper as a temporary solution here.
the work that WG component standard are doing is planned to solve the flags vs config problem, hopefully.
https://github.com/kubernetes/community/tree/master/wg-component-standard

one of the early steps in flight is a KEP for a fork of pflag (kflag):
https://github.com/kubernetes/enhancements/pull/764

Reply:
> This has actually made it harder to build tests on top. Instead of having a publicly exposed config struct that I can set any way I want and then use,

I don't understand the part about "harder to build tests". Previously it was impossible to build tests on top of the framework without modifying the framework. Now that part works.

> I must have my executable called to set these registered flags.

Yes, of course the executable must be called, because it is the one parsing flags and/or configuration. Can you explain how and where you want to set the flags without involving the test suite binary?

I'm not trying to defend Viper here.  I've never liked it myself (no error messages when spelling an option wrong!) and if there is a better solution, then I'm sure that would be welcome.

Reply:
For example, I could previously choose take the configuration any way I wanted and set a value like this: https://github.com/openshift/origin/blob/master/cmd/openshift-tests/openshift-tests.go#L251 and we cannot anymore.

Reply:
You can see a similar example for the upgrade targets here: https://github.com/openshift/origin/blob/master/cmd/openshift-tests/upgrade.go#L57-L58 .  This pull has remove the logical separation of configuration from how that configuration is specified in a way that appears to be inconsistent with the direction of the rest of the project.

See patterns like this https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L39 and https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/builder_flags.go#L106 which allow separation, flexibility, and composition.

Reply:
> set a value like this: https://github.com/openshift/origin/blob/master/cmd/openshift-tests/openshift-tests.go#L251

That sets the viper configuration file and then expects some code to do something with it. In 1.12, that code looked like this: https://github.com/kubernetes/kubernetes/blob/609a4b9323a752fc0e9a51d310cc836baa037a20/test/e2e/framework/test_context.go#L325-L332

In master, Viper support is no longer part of the framework itself. That's exactly the flexibility that you are talking about: instead of mandating one particular config parser, let the caller decide. To get a similar effect, you now need to do something like this: https://github.com/kubernetes/kubernetes/blob/ad4727440cc4a5be5c9714ffc0baefc11ce6c7f6/test/e2e/e2e_test.go#L59-L67

So the API has changed, but I think you can still do what you want to do.

> https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L39

That's exactly how I would also do it. I even wrote utility code that automatically registers struct fields as individual parameters:
https://github.com/kubernetes/kubernetes/blob/ad4727440cc4a5be5c9714ffc0baefc11ce6c7f6/test/e2e/framework/config/config.go#L17-L34

Reply:
You can see this doesn't compose, right?  I cannot import the `test/e2e/lifecycle` package without it claiming global flags.  Then, because the flag is treated as an internal global variable, it's impossible to set externally and embed the included tests.

This is substantively worse than the composition described above where it possible to create the options, then set  them based on a config of the composer's choosing, then execute.

Reply:
> I don't think this leads to clean composition. I cannot import the `test/e2e/lifecycle` package without it claiming global flags. Then, because the flag is treated as an internal global variable, it's impossible to set externally and embed the included tests.

turns out you're hitting the same problem with klog: https://github.com/kubernetes/kubernetes/issues/75403 .  This pull created the same problem with the tests.  They're not importable anymore

Reply:
>  I cannot import the test/e2e/lifecycle package without it claiming global flags.

You also can't import the `e2e/framework` either without it claiming global flags. That's actually worse, because one always gets the flags even when not using the corresponding tests. My proposal ("proposal" because not completely implemented) makes that better by only defining flags when actually importing the tests that use them.

Anyway, I thought that using `flag` and its default `FlagSet` was an acceptable mode of operation. That was fine for the Kubernetes E2E suite and no-one mentioned other usages where that might not be appropriate.

It's not hard to change. Instead of allowing tests to add to the default `FlagSet`, we define a `framework/config.FlagSet` that is to be used for test config options. Then the user of the framework can decide whether it wants to expose those as actual command line options or only wants to set them indirectly through some other mechanism, like `viperconfig`.

Reply:
> You also can't import the `e2e/framework` either without it claiming global flags. That's actually worse, because one always gets the flags even when not using the corresponding tests. My proposal ("proposal" because not completely implemented) makes that better by only defining flags when actually importing the tests that use them.

Before I had to explicitly call a method to register the flags (`RegisterCommonFlags` as I recall).  Now you're doing it on import.  Before I could import a package and choose how to use it, now I can't import the package.

Reply:
> Before I had to explicitly call a method to register the flags (RegisterCommonFlags as I recall). Now you're doing it on import.

Right, I had forgotten that aspect. So you are asking for tests to register their flags only when called explicitly, right?

What about the Ginkgo tests? They also get registered during init, unconditionally. It's perhaps less likely that one wants to import a test package without actually using the tests, but it could happen. Someone might want to have finer control over which tests get registered, based on command line flags. We actually do have this issue right now for CSI testing: we use the Kubernetes E2E suite, but run only 28 out of 6740 tests. Loading all those JUnit files with the Skip entries is noticable slow in Spyglass.

Let me see whether I can come up with a solution...